text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,458 @@
+
+```js
+Sentry.init();
+
+// There is always an isolation scope
+const initialIsolationScope = Sentry.getIsolationScope();
+initialIsolationScope.setTag("tag1", "aa");
+
+// Auto-instrumentation automatically forks an execution context with a new isolation scope
+app.get("/my-route", function () {
+  const isolationScope = Sentry.getIsolationScope();
+  isolationScope !== initialIsolationScope;
+  // Isolation scope inherits from previously current isolation scope
+  isolationScope.getTags(); // -> tag1: aa
+});
+```
+
+Note that in environments where you do not have multiple execution contexts (e.g. Browser JS), there is basically only a single isolation scope that is fundamentally the same as the global scope. In these environments, global & isolation scope can be used completely interchangeably. However, for consistency we still have these concepts everywhere - but users do not need to care about them, and we can reflect this in the SDK specific documentation.
+
+### How to document isolation scopes
+
+Isolation scopes will be documented in a new section under "Enriching Events > Per request data" (or similar, based on SDK language & needs), which will explain how to use `getIsolationScope()` and when/how data added there will be applied.
+
+Global methods to set data, like `Sentry.setTag()` or `Sentry.setUser()`, will set on the isolation scope. Only if you want to set on a different scope, you need to manually get it (e.g. `getGlobalScope().setTag()`).
+
+### When to create an isolation scope
+
+For most server-side SDKs, an isolation scope will be created for each request being processed.
+Roughly, it will equate to each time we currently fork a hub.
+
+### Examples for isolation scopes
+
+Example for instrumentation that we would write:
+
+```ts
+function wrapHttpServerRequest(original: Function): Function {
+  // Fork an execution context for this server request, that is isolated
+  return Sentry.withIsolatedScope((scope) => {
+    // anything in here will have the same isolated scope!
+    return original();
+  });
+}
+```
+
+Example for hooking into external auto-instrumentation (e.g. OpenTelemetry):
+
+```ts
+let onRequestHook: (span: Span) => void;
+
+// This method is not defined by us, but is some external code
+// Here just for demonstration purposes of how that may be implemented
+function otelWrapHttpServerRequest(original: Function): Function {
+  // Fork an execution context for this server request,
+  // but without isolating this!
+  return Sentry.withScope((scope) => {
+    onRequestHook(trace.getActiveSpan());
+    return original();
+  });
+}
+
+// This would be our custom sentry configuration
+onRequestHook = () => {
+  const scope = getCurrentScope();
+  scope.isolate(); // Add an isolation scope to the already forked scope
+};
+```
+
+## Applying scopes
+
+Scopes are applied in this order to events:
+
+```ts
+class Scope {
```

Comment:
So, clients will not be aware that there is a scope. The `Scope` class merges data into the event before passing it to the client. Did I get this correctly?

Reply:
that's the idea, this is also how it works today (at least in JS and I _think_ more or less everywhere?) - this RFC does not propose to change this from it is today, except that the special scope data is _also_ applied to the events!